File: sources/perl/05_alloc_churn/main.pl
# task 05 alloc_churn — expected output: 1274991808
# build: perl main.pl    run: perl main.pl
# note: a Perl scalar is a mutable byte string, so every iteration allocates a fresh 64-byte string;
#       the slots write keeps the buffer reachable and drops the buffer it replaces.
use strict;
use warnings;

my $total = 0;
my @slots;
$#slots = 255;

for (my $i = 0; $i < 10000000; $i++) {
    my $buf = "\0" x 64;
    substr($buf, 0, 1) = chr($i & 255);
    $total += ord($buf);
    $slots[$i & 255] = $buf;
}

print $total, "\n";

File: sources/lua/05_alloc_churn/main.lua
-- task 05 alloc_churn — expected output: 1274991808
-- build: none (interpreted)    run: lua main.lua (PUC Lua 5.4)    also runs on LuaJIT: luajit main.lua
-- Lua strings are immutable, so each iteration builds a fresh 64-byte string; the slots
-- table keeps it reachable and drops the buffer it replaces.

local slots = {}
local total = 0

for i = 0, 9999999 do
    local buf = string.char(i % 256) .. string.rep("\0", 63)
    total = total + string.byte(buf)
    slots[(i % 256) + 1] = buf
end

print(total)
